evidence/reactor.go

```diff
@@ -131,90 +133,50 @@ func (evR *Reactor) broadcastEvidenceRoutine(peer p2p.Peer) {
-			retry = !success
-		}
-
-		if retry {
-			time.Sleep(peerCatchupSleepIntervalMS * time.Millisecond)
+		peerState, ok := peer.Get(types.PeerStateKey).(PeerState)
+		if !ok {
+			// Peer does not have a state yet. We set it in the consensus reactor, but
+			// when we add peer in Switch, the order we call reactors#AddPeer is
+			// different every time due to us using a map. Sometimes other reactors
+			// will be initialized before the consensus reactor. We should wait a few
+			// milliseconds and retry.
+			time.Sleep(broadcastEvidenceIntervalS * time.Second)
 			continue
 		}
 
-		afterCh := time.After(time.Second * broadcastEvidenceIntervalS)
-		select {
-		case <-afterCh:
-			// start from the beginning every tick.
-			// TODO: only do this if we're at the end of the list!
-			next = nil
-		case <-next.NextWaitChan():
-			// see the start of the for loop for nil check
-			next = next.Next()
-		case <-peer.Quit():
-			return
-		case <-evR.Quit():
-			return
+		var evMsg ListMessage
+		for next != nil && len(evMsg.Evidence) < messageSize {
+			ev := next.Value.(types.Evidence)
+			// check to see if peer is behind, if so jump to next piece of evidence
+			if ev.Height() <= peerState.GetHeight() {
+				// Update should ensure that all evidence in the list is not expired
+				// hence we don't need to check it
+				evMsg.Evidence = append(evMsg.Evidence, ev)
+			}
+			// As evidence arrives in batches, wait for a period to see if there is
+			// more evidence to come, if not then finish the msg and send it to the peer.
+			// (10ms is the allocated time for each loop of AddEvidence)
+			afterCh := time.After(time.Millisecond * 10)
+			select {
+			case <-afterCh:
+				next = nil
+			case <-next.NextWaitChan():
+				next = next.Next()
+			case <-peer.Quit():
+				return
+			case <-evR.Quit():
+				return
+			}
 		}
-	}
-}
-
-// Returns the message to send the peer, or nil if the evidence is invalid for the peer.
-// If message is nil, return true if we should sleep and try again.
-func (evR Reactor) checkSendEvidenceMessage(
-	peer p2p.Peer,
-	ev types.Evidence,
-) (msg Message, retry bool) {
-
-	// make sure the peer is up to date
-	evHeight := ev.Height()
-	peerState, ok := peer.Get(types.PeerStateKey).(PeerState)
-	if !ok {
-		// Peer does not have a state yet. We set it in the consensus reactor, but
-		// when we add peer in Switch, the order we call reactors#AddPeer is
-		// different every time due to us using a map. Sometimes other reactors
-		// will be initialized before the consensus reactor. We should wait a few
-		// milliseconds and retry.
-		return nil, true
-	}
-
-	// NOTE: We only send evidence to peers where
-	// peerHeight - maxAge < evidenceHeight < peerHeight
-	// and
-	// lastBlockTime - maxDuration < evidenceTime
-	var (
-		peerHeight = peerState.GetHeight()
 
-		params = evR.evpool.State().ConsensusParams.Evidence
+		success := peer.Send(EvidenceChannel, cdc.MustMarshalBinaryBare(evMsg))
```

Comment:
Instead of waiting 1 second before sending again, I would like to implement something with exponential backoff say 10 times before the peer is dropped and the go routine finished. What do you think @melekes - is there something like this that already exists (this should be the duty of the p2p package) or should I just write something

Reply:
yeah, we should probably write a library exp retry. usually we just write it every time because it's a few lines of code. dropping a peer because you can't send it data does not sound like a good idea. send buffer may be full... and you are going to remove peer because you have a full buffer on your side?